### ISSUE TITLE
Attachment Manager hanging "Loading your Library"

### ISSUE THREAD

**Issue by markNZed** [mickstar/Zoo-For-Zotero#67]
Thanks for your work. I copied the Zotero storage into the Attachments Location Custom Path (about 400 attachments). Then the Attachment Manager page hangs with a message "Loading your library". It was working with a single attachment. I am wondering if it is hanging because of the number of attachments?

Is it intended to be used in this way? It could be great if I can synchronize a folder on my Android device with the cloud then point to this for the attachments. This could synchronize without Zoo for Zotero needing to manage the synchronization of the attachments. 

I did experiment with WebDAV but that did not seem to be stable for uploading changes etc.

**Comment by mickstar** [comment]
400 is not much. If it's still hanging after 1 minute then it's broken.

I'm not too sure as to why, your workflow should work.

**Comment by markNZed** [comment]
No error message, after a while I get a system popup asking to shut down the app or wait. Can we enable any logging from the app?

**Comment by mickstar** [comment]
Android has something called logcat enabled but it's not very user
friendly. The app logs to that

**Comment by mickstar** [comment]
Here is the code that handles the attachment manager https://www.github.com/mickstar/Zoo-For-Zotero/tree/master/app%2Fsrc%2Fmain%2Fjava%2Fcom%2Fmickstarify%2Fzooforzotero%2FAttachmentManager%2FAttachmentManagerModel.kt

Just skimming it the only error I can think of is some kind of file permission error but I doubt that's the case. I'm not too sure what's wrong.

Can you tell me if the attachments open from the library view?

**Comment by markNZed** [comment]
I have logcat running and see messages like:

04-16 13:19:20.451  9260  9260 D zotero  : checking if Stoller_2009_Phenomenology and the Poststructural Critique of Experience.pdf exists
04-16 13:19:21.601  9260  9260 D zotero  : checking if Stern_Wittgenstein’s Philosophical Investigations.pdf exists

Most of them are not being printed that fast. I'm guessing a couple if seconds per PDF. This would take quite some time even for 400. 

Maybe logcat slows things down a lot?

**Comment by markNZed** [comment]
Also messages like:
04-16 13:21:48.030  9260  9270 I fy.zooforzoter: Background concurrent copying GC freed 115587(5MB) AllocSpace objects, 0(0B) LOS objects, 52% free, 10MB/22MB, paused 98us total 125.702ms

**Comment by mickstar** [comment]
Logcat is real time. That would explain the issue, it's checking if the
attachments exist really slowly. Not sure how to speed that up given its
using the android storage framework API which is notoriously finicky.

Why do you need to use the attachment manager? The library view should work
given you have already downloaded the attachments

**Comment by mickstar** [comment]
You could fix it by using the external cache directory and copying the
attachments to that folder.

**Comment by markNZed** [comment]
OK. I will check if it does work. It is just so slow I thought things were broken. It will take a little while to finish the checks.

**Comment by markNZed** [comment]
The external cache directory is the Download dir? Can we change that?

**Comment by mickstar** [comment]
It's the path
/Android/data/com.mickstarify.zooforzotero/cache
That you will find in your storage path.

**Comment by markNZed** [comment]
After taking so long to check all the files the app still hung. It stopped printing messages about checking files. I guess things go wrong when a process takes so long.

**Comment by markNZed** [comment]
Maybe this is an idea: https://www.b4x.com/android/forum/threads/why-is-a-check-if-file-exists-so-slow.66313/

**Comment by markNZed** [comment]
If I put them in the cache then Attachment Manager returns almost immediately

**Comment by mickstar** [comment]
Yep that's the way it should be. There's something wrong with the custom
path option.

**Comment by markNZed** [comment]
It seems the file gets transferred to the Downloads directory when I open it. Maybe this is a "feature" of the PDF reader I'm using. Is there a way to change which PDF reader the app uses?

**Comment by markNZed** [comment]
Ah. I guess the PDF reader can't access the cache. So we do need the external storage so edits to the PDF can be kept in sync.

**Comment by mickstar** [comment]
Are you opening this PDFs via zoo?

**Comment by markNZed** [comment]
Yes, I tap on it and it opens. If I set the external storage then it opens the file as expected. So I guess I just avoid using the Attachment Manager tab and it should be OK for me.

**Comment by fabeit** [comment]
@markNZed I am on an ereader and have similar problems. If I use cache then the attachments are downloaded from the "attachment manager", if I set a custom path the attachment manager hangs and I need to force stop Zoo. 

Also when I open a pdf from the library view it is "downloaded" to the download folder, why is it not opened directly fro the cache directory?

**Comment by markNZed** [comment]
In the app Settings I have Attachments Location set to Custom Path and that is then pointing to a directory that is synced via Google Drive (using DriveSync Pro). On my desktop PC I have a link from the ~/Zotero/storage directory to the directory in the Google Drive folder. This seems to work fairly well - I can add comments in the PDF using the ereader and they appear on my desktop automatically (and vice-versa). I do not use the Attachment Manager in the app.

**Comment by fabeit** [comment]
@markNZed thanks for the tip, that works better but I'd rather use the webDAV for all my files. The only issue I have now si that Zoo doesn't offer to upload the file and I have to manually to do. Does Zoo automatically upload the PDFs you modify?

**Comment by markNZed** [comment]
Yes the PDFs are synced in both directions but in my setup it is Google Drive that syncs the PDFs and Zoo only syncs the metadata via webDAV. If you get webDAV working then you will run into the performance issues I mentioned above in the app.
